### DISCUSSION TITLE
Webpack의 Code Splitting에 대해 설명해 주세요

### DISCUSSION THREAD

**Discussion by jswith** [woowacourse-study/2022-woowahan-bansanghwe#72] [Optimization]
# Code Splitting in Webpack

코드 스플리팅이란 무엇일까요? 코드를 다양한 번들로 분할하고, 요청에 따라 로드하거나 병렬로 로드할 수 있도록 코드를 나누는 기능입니다. 번들을 더 작게 나누고 리소스 우선순위를 올바르게 제어하기 위해서 사용하며, 잘 활용하면 로드 시간을 단축할 수 있습니다. 어떤 컴퓨터가 1초에 10의 크기를 3개 로드할 수 있다고 가정하면, 100 크기의 파일 1개 로드하는 것(10초 소요)보다 파일을 나눠서 30, 30, 40 크기의 파일을 로드하는 것(4초 소요)이 더 빠르겠죠? 만일 lazy loading과 결합한다면 파일을 나눈 뒤 당장 필요한 파일만 불러올 수도 있습니다.

웹팩에서 코드 스플리팅은 세 가지 방식으로 접근할 수 있는데요. `Entry Points`, `Prevent Duplication`, `Dynamic imports`입니다.

- Entry Points: `entry` 설정을 사용하여 코드를 수동으로 분할합니다.
- Prevent Duplication: `Entry dependencies` 또는 `SplitChunksPlugin`을 사용하여 중복 청크를 제거하고 청크를 분할합니다.
- Dynamic imports: 모듈 내에서 인라인 함수 호출을 통해 코드를 분할합니다.

그럼, 이제부터 세 가지 방식을 하나씩 살펴봅시다.

## Entry Points

코드를 분할하는 가장 기본적이고 쉬운 방법입니다. Entry Points를 통해 메인 번들에서 다른 모듈을 어떻게 분리하는지 알아봅시다. 몇 가지 주의할 점도 있는데 이에 대해서도 다루어볼게요.

- 프로젝트 구조

  ```
  webpack-demo
  |- package.json
  |- package-lock.json
  |- webpack.config.js
  |- /dist
  |- /src
    |- index.js
    |- another-module.js
  |- /node_modules
  ```

- another-module.js

  ```javascript
  import _ from 'lodash';
  
  console.log(_.join(['Another', 'module', 'loaded!'], ' '));
  ```

- webpack.config.js

  ```javascript
  const path = require('path');
  
  module.exports = {
    mode: 'development',
    entry: {
      index: './src/index.js',
      another: './src/another-module.js',
    },
     output: {
       filename: '[name].bundle.js',
       path: path.resolve(__dirname, 'dist'),
     },
   };
  ```

entry를 index와 another로 분리하고, output의 filename을 `'[name]'.bundle.js`형태로 지정했습니다. 위 설정으로 빌드를 하게되면 `index.bundle.js`와 `another.bundle.js` 파일이 생성됩니다.

- 빌드 결과

```
...
[webpack-cli] Compilation finished
asset index.bundle.js 553 KiB [emitted] (name: index)
asset another.bundle.js 553 KiB [emitted] (name: another)
runtime modules 2.49 KiB 12 modules
cacheable modules 530 KiB
  ./src/index.js 257 bytes [built] [code generated]
  ./src/another-module.js 84 bytes [built] [code generated]
  ./node_modules/lodash/lodash.js 530 KiB [built] [code generated]
webpack 5.4.0 compiled successfully in 245 ms
```

그런데, 이 방식에는 몇 가지 주의할 점이 있는데요.

- 엔트리 청크 사이에 중복된 모듈이 있으면, 두 번들에 모두 포함이 됩니다.
  - 위 예제에서는 `./src/index.js`에서도 `lodash`를 가져오므로 `index.js`와  `another-moudle.js` 양쪽 번들 모두에 `lodash`가 중복으로 포함되는 문제가 발생합니다. 다음 섹션(Prevent Duplication)에서 중복을 제거하는 방법을 살펴보도록 하겠습니다.
- 코어 애플리케이션 로직의 코드를 동적으로 분할하는 데는 사용할 수 없습니다.

### Entry dependencies

dependOn 옵션을 사용하면 청크간 모듈을 공유할 수 있습니다. 즉 각 청크마다 모듈을 불러오는 것이 아니라, 하나의 모듈을 불러오고 여러 개의 청크가 공유해서 사용할 수 있는 것이죠.

- webpack.config.js

```javascript
const path = require('path');

module.exports = {
   mode: 'development',
   entry: {
    index: {
      import: './src/index.js',
      dependOn: 'shared',
    },
    another: {
      import: './src/another-module.js',
      dependOn: 'shared',
    },
    shared: 'lodash',
   },
   output: {
     filename: '[name].bundle.js',
     path: path.resolve(__dirname, 'dist'),
   },
 };
```

index와 another에 `dependOn: 'shared'` 옵션을 추가하고, `shared: 'lodash'` 옵션을 추가하면 index와 another 번들이 lodash를 공유하게 됩니다.

그런데 주의할 점이 있습니다. 단일 HTML 페이지에서 여러 엔트리 포인트를 사용하는 경우에는 또 다른 옵션이 필요한데요. 바로 `optimization.runtimeChunk: 'single'` 옵션입니다. 이 옵션이 왜 필요할까요? 옵션을 설정하지 않으면 모듈이 여러 번 인스턴스화되기 때문입니다. 

예시를 통해 살펴볼까요?

- component-1.js

```javascript
import obj from './obj.js';
obj.count++;
console.log('component-1', obj.count);
```

- component-2.js

```javascript
import obj from './obj.js';
obj.count++;
console.log('component-2', obj.count);
```

- obj.js

```javascript
export default { count: 0 };
```

`obj.js` 모듈이 번들에 각각 인스턴스화되면 각 번들은 자체 .count 속성을 증가시키고 그 결과는 각각 카운트 값이 1인 두 개의 개체가 됩니다.

그러나, 모듈이 한 번만 올바르게 인스턴스화되면 `component-1.js`, `component-2.js` 모두 `obj.js`에서 내보낸 동일한 개체에 대한 참조를 수신하고 해당 개체의 .count 속성이 두 번 증가하여 값이 2가 됩니다.

여기까지가  `optimization.runtimeChunk: 'single'`을 사용하는 이유에 대한 예시였습니다.

이제, `optimization.runtimeChunk: 'single'` 옵션을 추가하고 빌드를 해보도록 하겠습니다.

- webpack.config.js

```javascript
const path = require('path');

module.exports = {
   mode: 'development',
   entry: {
     index: {
       import: './src/index.js',
       dependOn: 'shared',
     },
     another: {
       import: './src/another-module.js',
       dependOn: 'shared',
     },
     shared: 'lodash',
   },
   output: {
     filename: '[name].bundle.js',
     path: path.resolve(__dirname, 'dist'),
   },
  optimization: {
    runtimeChunk: 'single',
  },
 };
```

- 빌드 결과

```
...
[webpack-cli] Compilation finished
asset shared.bundle.js 549 KiB [compared for emit] (name: shared)
asset runtime.bundle.js 7.79 KiB [compared for emit] (name: runtime)
asset index.bundle.js 1.77 KiB [compared for emit] (name: index)
asset another.bundle.js 1.65 KiB [compared for emit] (name: another)
Entrypoint index 1.77 KiB = index.bundle.js
Entrypoint another 1.65 KiB = another.bundle.js
Entrypoint shared 557 KiB = runtime.bundle.js 7.79 KiB shared.bundle.js 549 KiB
runtime modules 3.76 KiB 7 modules
cacheable modules 530 KiB
  ./node_modules/lodash/lodash.js 530 KiB [built] [code generated]
  ./src/another-module.js 84 bytes [built] [code generated]
  ./src/index.js 257 bytes [built] [code generated]
webpack 5.4.0 compiled successfully in 249 ms
```

빌드 결과, `shared.bundle.js`, `index.bundle.js` 및 `another.bundle.js` 외에 또 다른 `runtime.bundle.js` 파일이 생성됩니다.

### SplitChunksPlugin

SplitChunksPlugin은 기존 엔트리 청크 또는 완전히 새로운 청크로 공통 의존성을 추출할 수 있도록 해줍니다. 이를 활용하여 lodash 중복을 제거해볼까요? `optimiaztion.splitChunks.chunks: 'all'` 옵션을 적용합니다. 옵션을 추가하고 빌드를 해보겠습니다.

- webpack.config.js

```javascript
const path = require('path');

module.exports = {
    mode: 'development',
    entry: {
      index: './src/index.js',
      another: './src/another-module.js',
    },
    output: {
      filename: '[name].bundle.js',
      path: path.resolve(__dirname, 'dist'),
    },
   optimization: {
     splitChunks: {
       chunks: 'all',
     },
   },
  }
```

- 빌드 결과

```
...
[webpack-cli] Compilation finished
asset vendors-node_modules_lodash_lodash_js.bundle.js 549 KiB [compared for emit] (id hint: vendors)
asset index.bundle.js 8.92 KiB [compared for emit] (name: index)
asset another.bundle.js 8.8 KiB [compared for emit] (name: another)
Entrypoint index 558 KiB = vendors-node_modules_lodash_lodash_js.bundle.js 549 KiB index.bundle.js 8.92 KiB
Entrypoint another 558 KiB = vendors-node_modules_lodash_lodash_js.bundle.js 549 KiB another.bundle.js 8.8 KiB
runtime modules 7.64 KiB 14 modules
cacheable modules 530 KiB
  ./src/index.js 257 bytes [built] [code generated]
  ./src/another-module.js 84 bytes [built] [code generated]
  ./node_modules/lodash/lodash.js 530 KiB [built] [code generated]
webpack 5.4.0 compiled successfully in 241 ms
```

빌드를 하게 되면, `index.bundle.js`와 `another.bundle.js`에서 `lodash` 중복 의존성이 제거된 것을 확인 할 수 있습니다. `lodash`를 별도의 청크로 분리하기 때문입니다.

## Dynamic Imports

웹팩은 동적 코드 스플리팅에 두 가지 기술을 지원합니다. 첫 번째는 `import 구문`을 사용하는 방식이고, 또 다른 방식은 `require.ensure`를 사용하는 방식입니다. 첫 번째 방식을 살펴볼 예정인데요. 그 전에 웹팩 설정 파일을 수정하고, 프로젝트의 구조를 살짝 변경해보겠습니다.

- webpack.config.js

```javascript
const path = require('path');

 module.exports = {
   mode: 'development',
   entry: {
     index: './src/index.js',
   },
   output: {
     filename: '[name].bundle.js',
     path: path.resolve(__dirname, 'dist'),
   },
 };
```

- 프로젝트 구조

```
webpack-demo
|- package.json
|- package-lock.json
|- webpack.config.js
|- /dist
|- /src
  |- index.js
|- /node_modules
```

이제 정적으로 가져왔던 lodash를 동적으로 가져오고, 청크를 분리해보겠습니다.

- src/index.js

```javascript
function getComponent() {
  return import('lodash')
    .then(({ default: _ }) => {
      const element = document.createElement('div');
      element.innerHTML = _.join(['Hello', 'webpack'], ' ');

      return element;
    })
    .catch((error) => 'An error occurred while loading the component');
 }

getComponent().then((component) => {
  document.body.appendChild(component);
});
```

- 빌드 결과

```
...
[webpack-cli] Compilation finished
asset vendors-node_modules_lodash_lodash_js.bundle.js 549 KiB [compared for emit] (id hint: vendors)
asset index.bundle.js 13.5 KiB [compared for emit] (name: index)
runtime modules 7.37 KiB 11 modules
cacheable modules 530 KiB
  ./src/index.js 434 bytes [built] [code generated]
  ./node_modules/lodash/lodash.js 530 KiB [built] [code generated]
webpack 5.4.0 compiled successfully in 268 ms
```

빌드 결과 lodash가 별도의 번들로 분리된 것을 확인할 수 있습니다. `import()`는 promise를 반환하므로 async 함수와 함께 사용할 수 있습니다. 코드를 단순화해보겠습니다.

```javascript
async function getComponent() {
  const element = document.createElement('div');
  const { default: _ } = await import('lodash');

  element.innerHTML = _.join(['Hello', 'webpack'], ' ');
  return element;
 }

 getComponent().then((component) => {
   document.body.appendChild(component);
 });
```

`import()`에 `dynamic expression`을 사용하면, 계산된 변수를 기반으로 특정 모듈을 가져올 수 있는데요. 이 부분에 대해서는 나중에 포스팅보도록 하겠습니다.

## Prefetching/Preloading modules

Prefetching과 Preloading은 웹팩 4.6.0+에서 지원하는 기능입니다. 

모듈을 가져올 때 인라인 지시문을 사용하면 웹팩이 브라우저에 다음과 같은 힌트를 제공합니다.

- **prefetch** : 향후 일부 탐색에 리소스가 필요할 수 있다.
- **preload** : 현재 탐색 중에 리소스가 필요하다.

prefetch의 예제를 살펴볼까요? `HomePage` 컴포넌트에서 `LoginButton` 컴포넌트를 렌더링하고, 이 컴포넌트를 클릭하면 `LoginModal` 컴포넌트를 요청하여 로드하는 경우를 가정해보겠습니다.

- LoginButton.js

```javascript
//...
import(/* webpackPrefetch: true */ './path/to/LoginModal.js');
```

위 코드는 페이지의 head 태그에 `<link rel="prefetch" href="login-modal-chunk.js">`를 추가하고 브라우저에 `login-modal-chunk.js`를 유휴 시간에 미리 가져오도록 지시합니다.

그렇다면 preload와 prefetch는 어떤 차이점이 있을까요?

- preload 청크는 부모 청크와 병렬로 로드를 시작합니다. prefetch 청크는 부모 청크가 로드 완료된 후에 로드를 시작합니다.
- preload 청크는 중간 우선순위를 가지며 즉시 다운로드됩니다. prefetch 청크는 브라우저가 유휴 상태일 때 다운로드 됩니다.
- preload 청크는 부모 청크에서 즉시 요청되어야 합니다. prefetch 청크는 나중에 언제라도 사용할 수 있습니다.
- 지원하는 브라우저에 차이가 있습니다.

이제 preload의 예제를 살펴보도록 하겠습니다. 별도의 청크에 있어야 하는 큰 라이브러리에 항상 의존하는 `Component`를 생각해 볼 수 있습니다. 거대한 `ChartingLibrary`가 필요한 `ChartComponent`를 상상해 봅시다. 렌더링 될 때 `LoadingIndicator`를 표시하고 즉시 `ChartingLibrary`를 요청하여 불러옵니다.

- ChartComponent.js

```javascript
//...
import(/* webpackPreload: true */ 'ChartingLibrary');
```

`ChartComponent`를 사용하는 페이지를 요청할 때 `<link rel="preload">`를 통해 charting-library-chunk도 병렬로 요청됩니다. page-chunk가 더 작고 더 빨리 완료된다고 가정하면 이미 요청된 `charting-library-chunk`가 완료될 때까지 페이지에는 `LoadingIndicator`가 보일 겁니다. 이처럼 `webpackPreload`를 잘못 사용하면 로드 시간이 증가할 수 있고, 실제로 성능이 저하 될 수 있으므로 preload를 사용할 시에는 주의가 필요합니다.

## References

- https://webpack.js.org/guides/code-splitting/
- https://bundlers.tooling.report/code-splitting/multi-entry/

**Comment by JUDONGHYEOK** [comment]
코드스플리팅이란 무엇인가요?

**Comment by JUDONGHYEOK** [comment]
코드를 분할하는 방법은?

**Comment by JUDONGHYEOK** [comment]
entry 포인트 말고 다른것?

**Comment by JUDONGHYEOK** [comment]
코드 스플리팅을 사용해본 경험은?

**Comment by JUDONGHYEOK** [comment]
React.lazy 사용했을 때 단점

**Comment by JUDONGHYEOK** [comment]
코드스플리팅을 하는 민초만의 목적

**Comment by JUDONGHYEOK** [comment]
모듈이란?

**Comment by JUDONGHYEOK** [comment]
모듈의 장점?

**Comment by JUDONGHYEOK** [comment]
webpack 이전에서의 모듈 관리 방식?

**Comment by JUDONGHYEOK** [comment]
require 과 import/export 차이

**Comment by JUDONGHYEOK** [comment]
성능이 중요한 이유?

**Comment by JUDONGHYEOK** [comment]
페이지간에 레이지 로딩이 중요한가 코드 스플리팅이 중요한가?

**Comment by JUDONGHYEOK** [comment]
preload와 prefetch?

**Comment by JUDONGHYEOK** [comment]
어떤 리소스에 대해서 preload를 하는 것이 좋다고 생각하는가?

**Comment by JUDONGHYEOK** [comment]
컴포넌트란 무엇인지?

**Comment by JUDONGHYEOK** [comment]
react에서 component란?

**Comment by JUDONGHYEOK** [comment]
민초만의 component분리 기준?

**Comment by JUDONGHYEOK** [comment]
엘리먼트와 컴포넌트의 차이점?

**Comment by JUDONGHYEOK** [comment]
모든 모듈들을 chunks로 분리하지 않는이유?

**Comment by JUDONGHYEOK** [comment]
만약 민초가 생성할 chunk의 최소 크기(minsize)를 지정한다면 몇바이트로 지정하고 싶은지?
